app/models/category_user.rb

```diff
@@ -23,21 +19,43 @@ def self.watching_levels
   end
 
   def self.batch_set(user, level, category_ids)
-    records = CategoryUser.where(user: user, notification_level: notification_levels[level])
-    old_ids = records.pluck(:category_id)
-    changed = false
+    level_num = notification_levels[level]
+    category_ids = Category.where(id: category_ids).pluck(:id)
 
-    category_ids = Category.where('id in (?)', category_ids).pluck(:id)
+    changed = false
 
-    remove = (old_ids - category_ids)
-    if remove.present?
-      records.where('category_id in (?)', remove).destroy_all
-      changed = true
+    # Update pre-existing category users
+    unless category_ids.empty?
+      changed ||=
+        CategoryUser
+          .where(user_id: user.id, category_id: category_ids)
+          .where.not(notification_level: level_num)
+          .update_all(notification_level: level_num) > 0
     end
 
-    (category_ids - old_ids).each do |id|
-      CategoryUser.create!(user: user, category_id: id, notification_level: notification_levels[level])
-      changed = true
+    # Remove extraneous category users
+    changed ||=
+      CategoryUser
+        .where(user_id: user.id, notification_level: level_num)
+        .where.not(category_id: category_ids)
+        .delete_all() > 0
+
+    # Create new category users
+    unless category_ids.empty?
+      # TODO: rewrite this when we upgrade to rails 6
+      changed ||=
+        DB.exec(
+          "
```

Comment:
FYI, I think `<<~SQL` (multiline heredoc strings in ruby) work really well with `DB.exec`. This string is fine, but it's something to consider.